src/audits/utils.ts

```diff
@@ -36,94 +36,125 @@ export function audit(name: AuditName, fn: () => Promise<void>): Audit {
-      max: T extends number ? T : never,
-    ) => {
-      if (!(min <= actual && actual <= max)) {
-        throw `${name} ${actual} is not between ${min} and ${max}`;
-      }
-    },
-    toContain: (
-      expected: T extends Array<infer U> ? U : T extends string ? T : never,
-    ) => {
-      // @ts-expect-error types will match, otherwise never
-      // eslint-disable-next-line @typescript-eslint/no-explicit-any
-      if (!actual.includes(expected as any)) {
-        throw `${name} ${JSON.stringify(
-          actual,
-        )} does not contain ${JSON.stringify(expected)}`;
-      }
-    },
-    notToContain: (
-      expected: T extends Array<infer U> ? U : T extends string ? T : never,
-    ) => {
-      // @ts-expect-error types will match, otherwise never
-      // eslint-disable-next-line @typescript-eslint/no-explicit-any
-      if (actual.includes(expected as any)) {
-        throw `${name} ${JSON.stringify(actual)} contains ${JSON.stringify(
-          expected,
-        )}`;
-      }
-    },
-    toHaveProperty: (
-      // eslint-disable-next-line @typescript-eslint/no-explicit-any
-      prop: T extends Record<any, any> ? PropertyKey : never,
-    ) => {
-      // @ts-expect-error types will match, otherwise never
-      if (!(prop in actual)) {
-        throw `${name} ${JSON.stringify(
-          actual,
-        )} does not have a property '${String(prop)}'`;
-      }
-    },
-    notToHaveProperty: (
-      // eslint-disable-next-line @typescript-eslint/no-explicit-any
-      prop: T extends Record<any, any> ? PropertyKey : never,
-    ) => {
-      // @ts-expect-error types will match, otherwise never
-      if (prop in actual) {
-        throw `${name} ${JSON.stringify(actual)} does have a property '${String(
-          prop,
-        )}'`;
-      }
-    },
-  };
+export class AuditError {
+  /**
+   * Response from the server.
+   */
+  public response: Response;
+  /**
+   * Reason for the failing audit.
+   */
+  public reason: string;
+  constructor(response: Response, reason: string) {
+    this.response = response;
+    this.reason = reason;
+  }
 }
 
 /**
- * Parses the string as JSON and safely reports parsing issues for audits.
+ * Will throw an AuditError if the assertion on Response fails.
+ *
+ * All fatal problems will throw an instance of an Error.
  *
- * Assumes the parsed JSON will be an `ExecutionResult`.
+ * The name "ressert" is a wordplay combining "response" and "assert".
  *
  * @private
- * */
-export async function assertBodyAsExecutionResult(res: Response) {
-  const str = await res.text();
-  try {
-    return JSON.parse(str) as ExecutionResult;
-  } catch (err) {
-    throw `Response body is not valid JSON. Got ${JSON.stringify(str)}`;
-  }
+ */
+export function ressert(res: Response) {
+  return {
+    status: {
+      toBe(code: number) {
+        if (res.status !== code) {
+          throw new AuditError(res, `Response status code is not ${code}`);
+        }
```

Comment:
We filter some of the SHOULD results from this test suite based on the text of the `reason`, so this broke our usage here:
https://github.com/apollographql/apollo-server/blob/37b3b7fb57ea105f40776166c9532536fd3f053d/packages/integration-testsuite/src/httpSpecTests.ts#L65

@enisdenjo This change also dropped the actual error code that was seen in the message (i.e. previously we expected a 400 in this string), it might be nice to restore `res.status` to the string here?

Reply:
Is there any expectation that the `reason` is stable (is this a breaking change that should be fixed in a patch?) or should we assume these might change and be prepared to update our filtering logic?

Reply:
Thanks for reporting! Sorry if this was a breaking change, those were not my intentions.

> This change also dropped the actual error code that was seen in the message (i.e. previously we expected a 400 in this string), it might be nice to restore res.status to the string here?

Failing audits now have the complete response in the `AuditFail.response` object. Can this be helpful in your case?

> Is there any expectation that the reason is stable (is this a breaking change that should be fixed in a patch?) or should we assume these might change and be prepared to update our filtering logic?

The reason itself is not subject to versioning, hence not a breaking change. However, I understand you use case - do you have any suggestions on identifying tests without depending on the reason? Maybe give each test an ID uniquely representing the case?

Reply:
@enisdenjo That's actually a great mitigation, thanks for the suggestion!
https://github.com/apollographql/apollo-server/pull/7316/commits/31736b5ec74846a5f64e964f265b284ad80a25eb

If the name is also not expected to be stable than a unique ID would be great. I would really like to be more explicit about which warnings we're allowing in our implementation. IDs would let me just do `allowedWarningIDs.includes(audit.id)` or something to that effect.

I suppose we could be doing this now with the names as I've alluded to, but maybe that's better as a human-readable description of the test which might be tweaked / improved upon in the future.

Reply:
@trevor-scheer I'd love to hear your thoughts on #49, would that be a fitting enhancement?